### PR TITLE
Performance: Optimize the isEditedPostEmpty selector

### PR DESCRIPTION
**Pull request by youknowriad**
Refs #11782 

The `isEditedPostEmpty` is a selector called on each key stroke, so optimizing it has an impact on the feeling of smoothness while typing in long posts, in this PR I avoid using the slow `getBlocks` selector in the `isEditedPostEmpty` selector.

Noting that I'm making this optimization because when I tried merging a lot of pending performance PRs together, this selector was taking 1/2 of the key strike time in a long post.

In the current state of master, I'm noticing a `13%` improvement while typing but I expect this number to be way bigger once all the other PRs are merged.

### CODE REVIEW

**Review by tofumatt — APPROVED** [review]
Nice 👍

**Comment by aduth** [comment]
I'd expect this to introduce some issue, specifically toward the special handling provided through `getBlocksForSerialization` for posts which consist of only one empty (unmodified) paragraph block. I guess that's handled by the fact that allowing `getEditedPostContent` to run would itself generate `getBlocksForSerialization` and produce the empty (thus falsey) result?

**Comment by youknowriad** [comment]
> I'd expect this to introduce some issue, specifically toward the special handling provided through getBlocksForSerialization for posts which consist of only one empty (unmodified) paragraph block. I guess that's handled by the fact that allowing getEditedPostContent to run would itself generate getBlocksForSerialization and produce the empty (thus falsey) result?

To be honest, I forgot this use-case but yeah it seems that it would work properly because of the fallback. And I think it's a fine drawback because we want to optimize the call when its cost is high (when it has a lot of blocks) while in this case it only contains a single block so the fallback will still be fast.

File: packages/editor/src/store/selectors.js [hunk 1]
```diff
@@ -454,18 +454,18 @@ export function isEditedPostSaveable( state ) {
  * @return {boolean} Whether post has content.
  */
 export function isEditedPostEmpty( state ) {
-	const blocks = getBlocksForSerialization( state );
+	const rootBlocks = getBlockOrder( state );
```
Comment (aduth) [inline packages/editor/src/store/selectors.js]: It'd be more accurate to describe this as a set of client IDs, not blocks, i.e. `rootBlockClientIds`.

**Review by aduth — COMMENTED** [review]
I agree that I think in practice this should always work, based on the current logic of `getBlocksForSerialization`. That said, it very much ties this selector to that specific implementation. We should at least very clearly document this. An alternative might be to have some separate `getBlockClientIdsForSerialization` for the purpose of avoiding the call to `getBlocks` while still respecting the logic of a "for serialization" condition.

**Comment by youknowriad** [comment]
@aduth Feel free to update to document properly, I'm not certain I understand exactly what comment you want to add.

**Comment by aduth** [comment]
What are your thoughts about a `getBlockClientIdsForSerialization` selector? My main reservation is that, as implemented, we have a fragile assumption taking place in `isEditedPostEmpty`.

Pro `getBlockClientIdsForSerialization`:

- Not prone to same fragility
- Can be optimized nearly just as well, since `getBlocksForSerialization` only does anything "special" if there's a single block in the post, which can be determined by a trivial `getBlockOrder` function call prior to inspecting the full block object

Con `getBlockClientIdsForSerialization`:

- Getting into "excessively specific" selector territory
- Still has slightly more overhead than as considered here
- `isEditedPostEmpty` already does some bad practice by assuming internals, what's a bit more? 😄

**Comment by youknowriad** [comment]
> Getting into "excessively specific" selector territory

This is my concern. I'm not too worried about selectors assuming internals as selectors are very related to the state shape which is also an internal. That said I'm not strongly against `getBlockClientIdsForSerialization`. We could add it without exposing it but we'd still need a `getBlock` call to check `isUnmodifiedDefaultBlock` if the block count is 1, so not great either.

**Comment by aduth** [comment]
> but we'd still need a `getBlock` call to check `isUnmodifiedDefaultBlock` if the block count is 1, so not great either.

Yeah, the thought is that if there's only one block, we're nowhere near the edge case of degraded performance which this pull request primarily impacts.

I'm actually wondering if the optimization here could be aimed at `getBlocksForSerialization`, keeping everything else intact and just avoiding the `getBlocks` in favor of a more guarded condition to first verify trivially there's one block in content.

**Comment by aduth** [comment]
In adding some unit tests here, I found that the current logic actually fails to account for `getBlocksForSerialization`, specifically in that the following condition will wrongly return `false` when, for an unmodified default block, `isEditedPostEmpty` should return `true`:

https://github.com/WordPress/gutenberg/blob/fded8017da008d7c751c64fc7374389ddcf40207/packages/editor/src/store/selectors.js#L477-L479

**Review by aduth — APPROVED** [review]
Pushed some changes to fix the aforementioned issue, expand on inline comments, and add relevant unit tests.

Feel free to merge if it looks okay to you.

**Comment by youknowriad** [comment]
Thanks for the update Andrew, appreciate it. These changes look good to me.

**Pull request merged**